From c330fc1b5cea2d6786502e88e76e7674730c2d70 Mon Sep 17 00:00:00 2001
From: "Rywacki, Kamil" <kamil.rywacki@intel.com>
Date: Mon, 23 Nov 2020 15:07:48 +0100
Subject: [PATCH] 
 PONRTSYS-8200-support-upstream-data-flow-for-multiple-uni-ports

- fix handling of returned value when adding new PCE rules
- fix subblk search for multiple CTP
---
 drivers/net/datapath/pon_qos/pon_qos_tc_pce.c | 46 +++++++++++++++++++++------
 1 file changed, 37 insertions(+), 9 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c b/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
index e70e0bdc113a..68f1bf5eb2d7 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -85,6 +86,35 @@ static int get_subif_data(struct net_device *dev,
 	return 0;
 }
 
+static int
+subblk_id_find(struct net_device *dev,
+	       GSW_PCE_RuleRegion_t region,
+	       int portid, int subif)
+{
+	struct pce_node *p;
+
+	list_for_each_entry(p, &tc_pce_list, list) {
+		if (p->region == GSW_PCE_RULE_COMMMON &&
+		    p->region == region) {
+			netdev_dbg(dev, "%s: subblk_id found in COMMON region: %d\n",
+				   __func__, p->subblk_id);
+			return p->subblk_id;
+		}
+		if (p->region == GSW_PCE_RULE_CTP &&
+		    p->region == region &&
+		    p->portid == portid &&
+		    p->subif == subif) {
+			netdev_dbg(dev, "%s: subblk_id found in CTP region(%d|%d): %d\n",
+				   __func__, p->portid, p->subif, p->subblk_id);
+			return p->subblk_id;
+		}
+	}
+
+	netdev_dbg(dev, "%s: subblk_id not found\n",
+		   __func__);
+	return 0;
+}
+
 #define TC_PCE_SUBBLK_SIZE 32
 
 int pon_qos_pce_rule_create(struct net_device *dev,
@@ -94,7 +124,6 @@ int pon_qos_pce_rule_create(struct net_device *dev,
 {
 	struct dp_pce_blk_info *pce_blk_info;
 	struct pce_node *tc_pce = NULL;
-	struct pce_node *p;
 	int portid, subif;
 	u32 index;
 	s32 ret;
@@ -124,19 +153,18 @@ int pon_qos_pce_rule_create(struct net_device *dev,
 	 * take subblk_id from the first rule on the list. If there are no
 	 * rules in storage, leave subblk_id equal to 0, to make DP create new
 	 * subblock. */
-	list_for_each_entry(p, &tc_pce_list, list) {
-		if (pce_rule->region == p->region) {
-			pce_blk_info->info.subblk_id = p->subblk_id;
-		}
-	}
+	pce_blk_info->info.subblk_id = subblk_id_find(dev, pce_rule->region,
+						      portid, subif);
 
-	index = dp_pce_rule_add(pce_blk_info, pce_rule);
-	if (index < 0) {
+	ret = dp_pce_rule_add(pce_blk_info, pce_rule);
+	if (ret < 0) {
 		netdev_err(dev, "%s: dp_pce_rule_add failed err %d\n",
-			   __func__, index);
+			   __func__, ret);
 		kfree(pce_blk_info);
 		return -EINVAL;
 	}
+
+	index = (u32)ret;
 	netdev_dbg(dev, "%s: pce rule added with index %d\n", __func__, index);
 
 	tc_pce = kzalloc(sizeof(*tc_pce), GFP_KERNEL);
